src/cls/queue/cls_queue_src.cc

```diff
@@ -0,0 +1,527 @@
+  encode(encoded_len, bl_encoded_len);
+
+  bl.claim_append(bl_encoded_len);
+  bl.claim_append(bl_head);
+
+  int ret = cls_cxx_write2(hctx, 0, bl.length(), &bl, CEPH_OSD_OP_FLAG_FADVISE_WILLNEED);
+  if (ret < 0) {
+    CLS_LOG(5, "ERROR: write_queue_head: failed to write head \n");
+    return ret;
+  }
+  return 0;
+}
+
+int get_queue_head(cls_method_context_t hctx, cls_queue_head& head)
+{
+  uint64_t chunk_size = 1024, start_offset = 0;
+
+  bufferlist bl_head;
+  int ret = cls_cxx_read(hctx, start_offset, chunk_size, &bl_head);
+  if (ret < 0) {
+    CLS_LOG(5, "ERROR: get_queue_head: failed to read head \n");
+    return ret;
+  }
+
+  //Process the chunk of data read
+  auto it = bl_head.cbegin();
+  // Queue head start
+  uint16_t queue_head_start;
+  try {
+    decode(queue_head_start, it);
+  } catch (buffer::error& err) {
+    CLS_LOG(0, "ERROR: get_queue_head: failed to decode queue start \n");
+    return -EINVAL;
+  }
+  if (queue_head_start != QUEUE_HEAD_START) {
+    CLS_LOG(0, "ERROR: get_queue_head: invalid queue start \n");
+    return -EINVAL;
+  }
+
+  it.seek(sizeof(uint16_t));
+  uint64_t encoded_len;
+  try {
+    decode(encoded_len, it);
+  } catch (buffer::error& err) {
+    CLS_LOG(0, "ERROR: get_queue_head: failed to decode encoded head size \n");
+    return -EINVAL;
+  }
+
+  if (encoded_len > (chunk_size - sizeof(uint64_t))) {
+    chunk_size = (encoded_len - (chunk_size - sizeof(uint64_t)));
+    start_offset += sizeof(uint64_t);
+    bufferlist bl_remaining_head;
+    int ret = cls_cxx_read2(hctx, start_offset, chunk_size, &bl_remaining_head, CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL);
+    if (ret < 0) {
+      CLS_LOG(5, "ERROR: get_queue_head: failed to read remaining part of head \n");
+      return ret;
+    }
+    bl_head.claim_append(bl_remaining_head);
+  }
+
+  it = bl_head.cbegin();
+  it.seek(sizeof(uint16_t) + sizeof(uint64_t));
+  try {
+    decode(head, it);
+  } catch (buffer::error& err) {
+    CLS_LOG(0, "ERROR: get_queue_head: failed to decode head\n");
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+int init_queue(cls_method_context_t hctx, const cls_init_queue_op& op)
+{
+  //get head and its size
+  cls_queue_head head;
+  int ret = get_queue_head(hctx, head);
+
+  //head is already initialized
+  if (ret == 0) {
+    return -EEXIST;
+  }
+
+  if (ret < 0 && ret != -EINVAL) {
+    return ret;
+  }
+
+  head = std::move(op.head);
+
+  if (op.has_urgent_data) {
+    if (op.head_size == 0) {
+      head.max_head_size = QUEUE_HEAD_SIZE_4K;
+      head.tail.offset = head.front.offset = QUEUE_START_OFFSET_4K;
+    } else {
+      head.max_head_size = op.head_size;
+      head.tail.offset = head.front.offset = head.max_head_size;
+    }
+    head.max_urgent_data_size = head.max_head_size - (sizeof(head.max_head_size) +
+                              sizeof(head.max_urgent_data_size) + sizeof(head.queue_size) +
+                              sizeof(head.tail) + sizeof(head.front));
```

Comment:
> this calculation would depend on the serialized length of cls_queue_head, not just the sum of its member sizes. the ENCODE_START() macro itself is adding some extra bytes for version numbers (see include/encoding.h) that you'd need to account for also - both from cls_queue_head::encode() and cls_queue_marker::encode()
> 
> the serialized length of cls_queue_head may also grow over time as we add new members; the encode/decode logic can handle this, but the max_urgent_data_size may change as a result

I understand that I should have taken into account Queue head start and the encoded length of the queue also for calculating the max urgent data size. Are you saying that the max urgent data size may change if the encoded length of the queue changes with fields getting added? Will the serialized size of encoded length not be fixed to sizeof(uint64_t) always?

Reply:
I think I understand what you mean when you say that the serialized size of queue will change when new fields will be added - so when that happens, we can update init_queue to update the queue head with the new fields and re-calculate the max urgent data entries size? Correct me if I am wrong. That brings me to another question - what if while updating the queue head with newer entries we find that the head doesn't have enough space (filled up by urgent data)

Reply:
> That brings me to another question - what if while updating the queue head with newer entries we find that the head doesn't have enough space (filled up by urgent data)

yeah, that's the tricky part. it seems like we would need some xattr-spillover logic every time we wrote the head, even if we weren't adding to urgent data

instead of letting the client specify a `head_size` on queue creation (and deriving the `max_urgent_data_size` from that), maybe we should do it the other way around? i.e. the client gives us a `max_urgent_data_size`, and we set `max_head_size = QUEUE_HEAD_SIZE_1K + max_urgent_data_size`. that way `max_urgent_data_size` stays constant, and cls_queue_head still has some room to grow

Reply:
Ok, so there are two things here:
1. max_urgent_entries which is exposed to the client as in cls_queue_client.cc, or cls_rgw_gc_client.cc.
2. Based on max_num_urgent_entries (which accounts for both the entries in head and xattrs), the client - cls_rgw_gc.cc will decide the max_urgent_data_size in terms of bytes. This will be a part of the metadata stored in the queue.
3. In cls_queue_src.cc we make the queue head of size 1K, irrespective of whether or not the client needs urgent data and then at the 1K boundary we write the urgent data which is bounded by max_urgent_data_size. In scenarios where urgent data size exceeds max_urgent_data_size, it spills over to xattrs.
This makes it easy for the queue head to grow till 1K (If we need to add entries that go beyond 1K in future, can we just spill those also over to xattrs?)

Is my understanding correct?

Reply:
> 3. In cls_queue_src.cc we make the queue head of size 1K, irrespective of whether or not the client needs urgent data and then at the 1K boundary we write the urgent data which is bound by max_urgent_data_size. In scenarios where urgent data size exceeds max_urgent_data_size, it spills over to xattrs.

writing the urgent data separately at offset=1k is an option, but i'd prefer to keep it packed within the cls_queue_head so that we can still fetch the head and urgent data in a single read op. the `max_head_size` (which represents where actual queue data starts) would be the same either way

Reply:
When you say that max_queue_head_size = QUEUE_HEAD_SIZE_1K + max_urgent_data_size, then QUEUE_HEAD_SIZE_1K is there for all fields before bl_urgent_data in the queue head right?
Also when we want to add new fields to the queue head, those will be part of QUEUE_HEAD_SIZE_1K?

Reply:
exactly right. as long as we enforce the max_urgent_data_size on cls_queue_head::bl_urgent_data, then the QUEUE_HEAD_SIZE_1K leaves wiggle room for the rest of cls_queue_head